ega-data-api-dataedge/src/main/java/eu/elixir/ega/ebi/dataedge/service/internal/EBINuFileService.java

```diff
@@ -0,0 +1,194 @@
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URISyntaxException;
+import java.security.InvalidAlgorithmParameterException;
+import java.security.InvalidKeyException;
+import java.security.NoSuchAlgorithmException;
+import java.util.concurrent.TimeUnit;
+
+public class EBINuFileService implements NuFileService {
+
+    private final KeyService keyService;
+    private final FileInfoService fileInfoService;
+    private final IFireService fireClientService;
+    private final LoadingCache<String, byte[]> aesHeaderCache;
+
+    public EBINuFileService(KeyService keyService, FileInfoService fileInfoService, IFireService fireClientService) {
+        this.keyService = keyService;
+        this.fileInfoService = fileInfoService;
+        this.fireClientService = fireClientService;
+        this.aesHeaderCache = Caffeine.newBuilder()
+                .expireAfterAccess(1, TimeUnit.DAYS)
+                .build(this::getAESHeader);
+    }
+
+    @Override
+    public long getPlainFileSize(String fileId) throws EgaFileNotFoundException, UnretrievableFileException, FileNotAvailableException {
+
+        return getPlainFileSize(getFile(fileId), keyService.getEncryptionAlgorithm(fileId));
+    }
+
+    private File getFile(String fileId) throws EgaFileNotFoundException, FileNotAvailableException {
+        File file;
+        try {
+            file = fileInfoService.getFileInfo(fileId);
+        } catch (NotFoundException exception) {
+            throw new EgaFileNotFoundException(fileId);
+        }
+
+        if (!file.getFileStatus().equals("available"))
+            throw new FileNotAvailableException(fileId);
+
+        return file;
+    }
+
+    private long getPlainFileSize(File file, String encryptionFormat) throws UnretrievableFileException {
+        long size = file.getFileSize();
+
+        if (encryptionFormat.equals("aes128") || encryptionFormat.equals("aes256")) {
+            size -= 16L;
+        } else if (!encryptionFormat.equals("plain")) {
+            throw new UnretrievableFileException(file.getFileId());
+        }
+
+        return size;
+    }
+
+    @Override
+    public InputStream getSpecificByteRange(String fileId, long start, long end) throws EgaFileNotFoundException, UnretrievableFileException, FileNotAvailableException, RangesNotSatisfiableException {
+        File file = getFile(fileId);
+        String encryptionAlgorithm = keyService.getEncryptionAlgorithm(fileId);
+
+        long plainFileSize = getPlainFileSize(file, encryptionAlgorithm);
+        if (start < 0 || end >= plainFileSize || end < start)
+            throw new RangesNotSatisfiableException(fileId, start, end);
+
+        try {
+            switch (encryptionAlgorithm) {
+                case "plain": {
+                    return fireClientService.downloadByteRangeByPath(file.getDisplayFilePath(), start, end);
+                }
+                case "aes128":
+                case "aes256": {
```

Comment:
plain, aes128 & aes256 can be defined as constants or enum & same can be used wherever required.